tesseract_collision/core/include/tesseract_collision/core/contact_managers_plugin_factory.h

```diff
@@ -256,12 +261,46 @@ class ContactManagersPluginFactory
    */
   YAML::Node getConfig() const;
 
+  /**
+   * @brief Set a global discrete contact manager factory create callback
+   *
+   * Callback function is called before plugins to allow explicit
+   * allocation of contact manager factory
+   *
+   * @param fn Contact manager factory create callback function
+   */
+  static void
+  setGlobalCreateDiscreteContactManagerFactoryCallback(const CreateDiscreteContactManagerFactoryCallbackFn& fn);
+
```

Comment:
Could we accomplish this by adding methods to the class like the one below? We have been going through and removing the use of function pointers to better support serialization.

``` c++ 
/**
   * @brief Add a discrete contact manager plugin
   * @param name The name
   * @param plugin The plugin
   */
  void addDiscreteContactManagerPlugin(const std::string& name, DiscreteContactManagerFactory::UPtr plugin);
```

Reply:
These are static functions so they shouldn't be serialized. These should be thought of as implementation details that are invisible to the user and only exist for specific use cases like wheels.

Reply:
Would have the proposed method serve the same need so the python wheel  includes the plugin libraries, or does it need to be static?

Reply:
My current concern is now there are two maps that could contain the same factories which one takes priority. 

Is it possible to create python specific function that takes a factory and adds the default factories or when the object is initialized in python add the defaults using the following?

``` c++

addDiscreteContactManagerPlugin("BulletDiscreteBVHManagerFactory", std::make_shared<tesseract_collision::tesseract_collision_bullet::BulletDiscreteBVHManagerFactory>());

addDiscreteContactManagerPlugin("BulletDiscreteSimpleManagerFactory", std::make_shared<tesseract_collision::tesseract_collision_bullet::BulletDiscreteSimpleManagerFactory>());

```

Reply:
Python doesn't know when these objects are created since they are created in various places in other libraries.

Reply:
We could add a single static callback to the plug-in loader and intercept requests there.

Reply:
Sorry for all of the questions, I just want to make sure I understand things. Is this mainly so that the python wheel pulls in the plugins? If so, if the libraries were included would the current plugin loading approach find them and load them?

Reply:
I tried linking in the plugins and using the loader, but the auditwheel process seems to destroy that functionality. This is intentional so library collisions don't happen.

Reply:
Was [this](https://stackoverflow.com/questions/19663366/dlopen-fails-in-boost-python-module) the potential issue?

Reply:
Do you have the python wheel that you tried linking in the plugins?

Reply:
That may be part of it. In general you don't want to load symbols into Python because it can break things. Typically the solution was to use static libraries, but auditwheel uses some tricks to make it work by mangling shared libraries.

Reply:
Here is the pull request with the Python changes: https://github.com/tesseract-robotics/tesseract_python/pull/23

The callbacks are configured here:

https://github.com/johnwason/tesseract_python-1/blob/03c1e067925c8d9de7bbba01d2ca0c0682fe6c0e/tesseract_python/swig/tesseract_collision_python.i#L71

https://github.com/johnwason/tesseract_python-1/blob/03c1e067925c8d9de7bbba01d2ca0c0682fe6c0e/tesseract_python/swig/tesseract_kinematics_python.i#L90

Reply:
I am not sure. If the callbacks return nullptr the loader will fall back to the normal library loader. I don't know if the libraries will load since all the other libraries are mangled.

For users that need plugins another method like conda or apt makes more sense, but most people using python won't need more than the built in capabilities.